### PR TITLE
Divisibility check and a stable merge

### PR DESCRIPTION
**Pull request by smr97**
This PR has two major changes, one is the `part_completed()` which serves as an abstract method that governs the division inside the adaptive scheduler. I see that in the merge, we might not want to divide the two iterators in several corner cases:
1. If both of them are sorted and one is strictly before or after another.
2. If both of them have the exact same number repeating (this is same as above).
3. *At least* one of them is "too small".
I am not sure if this list is exhaustive or if 3. is well defined. This means that we should have a feature in the API to stop the stealer from getting anything even if we are not done.

This PR addresses that. Furthermore, it has a waaayy more tedious merge that should now be stable. I have commented on the flow and hopefully taken care of all wierd corner cases. My feeling is that it is going to be slow.

A side note, the current stability test is weak. It would pass without the new merge, but just because of the change to part_completed. I will add a randomised one and maybe try to think of some corner cases also...

### CODE REVIEW

File: benches/merge.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            b.iter_with_setup(
+                || {
+                    let thread_pool = rayon::ThreadPoolBuilder::new()
+                        .num_threads(1)
+                        .build()
+                        .expect("Thread pool creation failed!");
+                    (thread_pool, interleaved_input(*input_size))
+                },
+                |(tp, (left, right, mut output))| {
+                    // TODO: I wonder if the install is not taking time here
```
Comment (smr97) [inline benches/merge.rs]: @wagnerf42  Good point! This one seems to be a sequential overhead comparison benchmark. For this, you can setup the global pool just once (maybe in the setup function). However, in my bench, I need to look closely, since I will also use the bindable thread pool, and it's install.

**Comment by smr97** [comment]
@wagnerf42 stability seems to be fixed now. Let me know what you think!
